src/osx/webrequest_urlsession.mm

```diff
@@ -0,0 +1,275 @@
+#endif
+
+@interface wxWebSessionDelegte : NSObject <NSURLSessionDataDelegate>
+{
+    wxWebSessionURLSession* m_session;
+    NSMapTable* m_requests;
+}
+
+@end
+
+@implementation wxWebSessionDelegte
+
+- initWithSession:(wxWebSessionURLSession*)session
+{
+    m_session = session;
+    m_requests = [[NSMapTable weakToStrongObjectsMapTable] retain];
+    return self;
+}
+
+- (void)dealloc
+{
+    [m_requests release];
+    [super dealloc];
+}
+
+- (void)registerRequest:(wxWebRequestURLSession*)request task:(NSURLSessionTask*)task
+{
+    [m_requests setObject:[NSValue valueWithPointer:request] forKey:task];
+}
+
+- (wxWebRequestURLSession*)requestForTask:(NSURLSessionTask*)task
+{
+    wxWebRequestURLSession* request = NULL;
+    NSValue* val = [m_requests objectForKey:task];
+    if (val)
+        request = static_cast<wxWebRequestURLSession*>(val.pointerValue);
+
+    return request;
+}
+
+- (void)URLSession:(NSURLSession *)session dataTask:(NSURLSessionDataTask *)dataTask didReceiveData:(NSData *)data
+{
+    wxWebRequestURLSession* request = [self requestForTask:dataTask];
+    if (request)
+        static_cast<wxWebResponseURLSession*>(request->GetResponse())->HandleData(data);
+}
+
+- (void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error
+{
+    wxWebRequestURLSession* request = [self requestForTask:task];
+    if (error)
+        request->SetState(wxWebRequest::State_Failed, wxCFStringRef(error.localizedDescription).AsString());
+    else
+        request->HandleCompletion();
+
+    // After the task is completed it no longer needs to be mapped
+    [m_requests removeObjectForKey:task];
+}
+
+@end
+
+//
+// wxWebRequestURLSession
+//
+wxWebRequestURLSession::wxWebRequestURLSession(wxWebSessionURLSession& session, const wxString& url, int id):
+    wxWebRequest(session, id),
+    m_url(url)
+{
+    m_headers = session.GetHeaders();
+}
+
+wxWebRequestURLSession::~wxWebRequestURLSession()
+{
+    [m_task release];
+}
+
+void wxWebRequestURLSession::Start()
+{
+    wxWebSessionURLSession& session = static_cast<wxWebSessionURLSession&>(GetSession());
+
+    NSMutableURLRequest* req = [NSMutableURLRequest requestWithURL:
+                                [NSURL URLWithString:wxCFStringRef(m_url).AsNSString()]];
+    if (m_method.empty())
+        req.HTTPMethod = (m_dataSize) ? @"POST" : @"GET";
+    else
+        req.HTTPMethod = wxCFStringRef(m_method).AsNSString();
+
+    // Set request headers
+    for (wxWebRequestHeaderMap::const_iterator it = m_headers.begin(); it != m_headers.end(); ++it)
+    {
+        [req setValue:wxCFStringRef(it->second).AsNSString() forHTTPHeaderField:
+         wxCFStringRef(it->first).AsNSString()];
+    }
+
+    if (m_dataSize)
+    {
+        // Read all upload data to memory buffer
+        wxMemoryBuffer memBuf;
+        m_dataStream->Read(memBuf.GetWriteBuf(m_dataSize), m_dataSize);
+
+        // Create NSDAta from memory buffer
```

Comment:
```suggestion
        // Create NSData from memory buffer
```
This looks rather inefficient, is there no way to write directly into `NSData` buffer?

Reply:
It looks like we can use `+ dataWithBytesNoCopy:length:` here, so I've changed the code do it like this.